compiler/noirc_frontend/src/hir/type_check/mod.rs

```diff
@@ -36,7 +37,7 @@ pub struct TypeChecker<'interner> {
     /// verified at the end of a function. This is because constraints arise
     /// on each variable, but it is only until function calls when the types
     /// needed for the trait constraint may become known.
-    trait_constraints: Vec<(TraitConstraint, ExprId)>,
+    trait_constraints: HashMap<ExprId, Vec<TraitConstraint>>,
```

Comment:
I also switched from a vec of all trait constraints to a hash map of ExprId -> Vec. I originally did it as I thought it may be necessary but I realized it was not. I felt this was a clearer association between expressions and trait constraints so I kept it. However, I can switch it back if we want as the vec is simpler